### PR TITLE
Optimising IO Stats Computation

### PR DESCRIPTION
**Pull request by Thejas-bhat**
The current way of IO Stats computation is causing too much of overhead:
- In case of _very high_ hit cardinality, the current implementation results in too many atomic.AddUint64 operations on an index level stat and that too its potentially invoked by a large number of independent queries running in the system (which are running as parallel threads).
- This essentially causes the stats updation to move from a per hit to a per query level (at the end of collector phase i.e.). So, DocumentMatch is used to track the bytes read information which will be aggregated in the collector level and the updating the same in the stats map using a callback.
- Furthermore, adding an extra interface "DiskStatsReporter" and introducing type assertions regarding that for the implementing struct causes additional overhead as well. So, the existing PostingsIterator, PostingsList and Segment interfaces now have DiskStatsReporter as an additional requirement to supported thereby avoiding the interface assertions altogether.

The issue which mentions the regression:
https://github.com/blevesearch/bleve/issues/1731

### CODE REVIEW

File: index/scorch/segment_plugin.go [hunk 1]
```diff
@@ -94,6 +94,20 @@ func RegisterSegmentPlugin(plugin SegmentPlugin, makeDefault bool) {
 	}
 }
 
+// This exported method allows the application
+// to set the state as to whether it needs
+// disk stats to be tracked while indexing and querying
+//
+// The stats said above refer to num_bytes_read_at_query_time
+// and num_bytes_written_at_index_time
+func SetDiskStatsCollection(state bool) {
```
Comment (Thejas-bhat) [inline index/scorch/segment_plugin.go:103]: Must be made made part of scorch options rather than an exported method.

File: index/scorch/segment_plugin.go [hunk 2]
```diff
@@ -94,6 +94,20 @@ func RegisterSegmentPlugin(plugin SegmentPlugin, makeDefault bool) {
...
+	zapv15.AccountIOStats = state
+}
+
+func collectDiskStats() bool {
```
Comment (Thejas-bhat) [inline index/scorch/segment_plugin.go:107]: collectDiskStats() -> accountIOStats()

**Review by Thejas-bhat — COMMENTED** [review]
Suggestions

File: index/scorch/snapshot_index.go [hunk 3]
```diff
@@ -149,7 +149,11 @@ func (i *IndexSnapshot) newIndexSnapshotFieldDict(field string,
 	for index, segment := range i.segment {
 		go func(index int, segment *SegmentSnapshot) {
 			var prevBytesRead uint64
-			seg, diskStatsAvailable := segment.segment.(diskStatsReporter)
+			var seg diskStatsReporter
+			var diskStatsAvailable bool
+			if accountIOStats() {
+				seg, diskStatsAvailable = segment.segment.(diskStatsReporter)
+			}
 			if diskStatsAvailable {
```
Comment (abhinavdangeti) [inline index/scorch/snapshot_index.go:157]: Move this snippet into within accountIOStats() ?

File: index/scorch/snapshot_index.go [hunk 4]
```diff
@@ -436,7 +440,11 @@ func (i *IndexSnapshot) Document(id string) (rv index.Document, err error) {
 
 	rvd := document.NewDocument(id)
 	var prevBytesRead uint64
-	seg, diskStatsAvailable := i.segment[segmentIndex].segment.(segment.DiskStatsReporter)
+	var seg segment.DiskStatsReporter
+	var diskStatsAvailable bool
+	if accountIOStats() {
+		seg, diskStatsAvailable = i.segment[segmentIndex].segment.(segment.DiskStatsReporter)
+	}
 	if diskStatsAvailable {
```
Comment (abhinavdangeti) [inline index/scorch/snapshot_index.go:448]: ditto^ and also everywhere else.

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]
You can update the zap v15 version based on the SHA for now.
We can set up formal release tags for zap mid next week.

File: index/scorch/reader_test.go [hunk 5]
```diff
@@ -153,6 +153,7 @@ func TestIndexReader(t *testing.T) {
 	if err != nil {
 		t.Errorf("unexpected error: %v", err)
 	}
+	match.BytesRead = 0
```
Comment (abhinavdangeti) [inline index/scorch/reader_test.go:156]: Why is this step necessary?

File: index/scorch/snapshot_index.go [hunk 6]
```diff
@@ -436,7 +431,9 @@ func (i *IndexSnapshot) Document(id string) (rv index.Document, err error) {
 
 	rvd := document.NewDocument(id)
 	var prevBytesRead uint64
-	seg, diskStatsAvailable := i.segment[segmentIndex].segment.(segment.DiskStatsReporter)
+	var seg segment.DiskStatsReporter
+	var diskStatsAvailable bool
```
Comment (abhinavdangeti) [inline index/scorch/snapshot_index.go:435]: This following type assertion is now not needed yes?

File: index/scorch/snapshot_index.go [hunk 7]
```diff
@@ -471,6 +468,7 @@ func (i *IndexSnapshot) Document(id string) (rv index.Document, err error) {
 	if err != nil {
 		return nil, err
 	}
+
 	if diskStatsAvailable {
```
Comment (abhinavdangeti) [inline index/scorch/snapshot_index.go:472]: Flag not necessary?

File: test/versus_test.go [hunk 8]
```diff
@@ -343,12 +343,15 @@ func testVersusSearches(vt *VersusTest, searchTemplates []string, idxA, idxB ble
 				if len(hitB.FieldTermLocations) == 0 {
 					hitB.FieldTermLocations = nil
 				}
+				// hitA.BytesRead = 0
+				// hitB.BytesRead = 0
```
Comment (abhinavdangeti) [inline test/versus_test.go:347]: Drop these comments.

File: test/versus_test.go [hunk 9]
```diff
@@ -386,7 +389,7 @@ func hitsById(res *bleve.SearchResult) map[string]*search.DocumentMatch {
 		hit.Score = math.Trunc(hit.Score*1000.0) / 1000.0
 		hit.IndexInternalID = nil
 		hit.HitNumber = 0
-
+		hit.BytesRead = 0
```
Comment (abhinavdangeti) [inline test/versus_test.go:392]: Again, why is this necessary?

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]

File: test/versus_test.go (hunk 9 above)
Comment (Thejas-bhat) [reply test/versus_test.go:392]: BytesRead doesn't relevance to that particular test, so I've disabled comparison over that. I've included comments for such settings.

File: index/scorch/snapshot_index.go (hunk 7 above)
Comment (Thejas-bhat) [reply index/scorch/snapshot_index.go]: Yeah these were minor rebase discrepancies. I've fixed them
